sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -70,437 +75,362 @@ object AggregateUtils {
+            // but at this point it is either risk an OOM/cudf error and potentially work or
+            // not work at all.
+            logWarning(s"Unable to merge reduction-only aggregated batches within " +
+                s"target batch limit of ${targetBatchSize}, attempting to merge remaining " +
+                s"${aggregatedBatches.size()} batches beyond limit")
+            withResource(mutable.ArrayBuffer[LazySpillableColumnarBatch]()) { batchesToConcat =>
+              aggregatedBatches.forEach(b => batchesToConcat += b)
+              aggregatedBatches.clear()
+              val batch = concatenateAndMerge(batchesToConcat)
+              // batch does not need to be marked spillable since it is the last and only batch
+              // and will be immediately retrieved on the next() call.
+              aggregatedBatches.add(batch)
+            }
           }
+          return
         }
       }
     }
   }
-}
 
-/**
- * GpuHashAggregateExec - is the GPU version of HashAggregateExec, with some major differences:
- * - it doesn't support spilling to disk
- * - it doesn't support strings in the grouping key
- * - it doesn't support count(col1, col2, ..., colN)
- * - it doesn't support distinct
- * @param requiredChildDistributionExpressions this is unchanged by the GPU. It is used in
- *                                             EnsureRequirements to be able to add shuffle nodes
- * @param groupingExpressions The expressions that, when applied to the input batch, return the
- *                            grouping key
- * @param aggregateExpressions The GpuAggregateExpression instances for this node
- * @param aggregateAttributes References to each GpuAggregateExpression (attribute references)
- * @param resultExpressions the expected output expression of this hash aggregate (which this
- *                          node should project)
- * @param child incoming plan (where we get input columns from)
- */
-case class GpuHashAggregateExec(
-    requiredChildDistributionExpressions: Option[Seq[Expression]],
-    groupingExpressions: Seq[Expression],
-    aggregateExpressions: Seq[GpuAggregateExpression],
-    aggregateAttributes: Seq[Attribute],
-    resultExpressions: Seq[NamedExpression],
-    child: SparkPlan) extends UnaryExecNode with GpuExec with Arm {
+  /**
+   * Perform a single pass over the aggregated batches attempting to merge adjacent batches.
+   * @return true if at least one merge operation occurred
+   */
+  private def mergePass(): Boolean = {
+    val batchesToConcat: mutable.ArrayBuffer[LazySpillableColumnarBatch] = mutable.ArrayBuffer.empty
+    var wasBatchMerged = false
+    var concatSize: Long = 0L
+    var batchesLeftInPass = aggregatedBatches.size()
+
+    while (batchesLeftInPass > 0) {
+      closeOnExcept(batchesToConcat) { _ =>
+        var isConcatSearchFinished = false
+        // Although tempting to allow the pass to "wrap around" and pick up batches previously
```

Comment:
This is great.  Might be good to mention that this is because of first/last.  It would also be great if we could have a follow on issue for stable sort.  This is another place where that would be helpful and when I looked I didn't see an issue for it. I guess I forgot to file one.

Reply:
Filed #2884 to track this.